src/framework/mpas_block_decomp.F

```diff
@@ -285,8 +322,239 @@ subroutine mpas_block_decomp_cells_for_proc(dminfo, partial_global_graph_info, l
+      integer, dimension(:), allocatable :: local_cell_list
+      integer, dimension(:), allocatable :: local_block_list
+
+      integer, dimension(:), allocatable ::global_block_id_arr, local_block_id_arr
+      integer :: i, global_block_id, local_block_id, iunit, ounit, istatus, ostatus, j, k
+      integer :: err, ierr
+      integer, dimension(:), pointer :: local_nvertices
+      integer :: num_local_vertices !< Number of local vertices for this processor
+      character (len=StrKIND) :: filename, msg
+
+      integer :: nLocEdgesGraph = 0, nLocVerticesGraph = 0, edgelocsiz = 0, npart = 1
+      character (len=StrKIND) :: partitionFilePrefix
+      integer, dimension(:), allocatable :: edgeloctab, vertloctab
+      doubleprecision :: stradat (scotch_stratdim)
+      doubleprecision :: scotchgraph (SCOTCH_GRAPHDIM)
+      doubleprecision :: scotchdgraph (SCOTCH_DGRAPHDIM)
+      doubleprecision :: scotchdgraph_redist (SCOTCH_DGRAPHDIM)
+      integer :: mpi_ierr
+#ifdef MPAS_USE_MPI_F08
+      type (MPI_Comm) :: localcomm
+#else
+      integer :: localcomm 
+#endif
+      
+      allocate(local_nvertices(dminfo % nprocs))
+      allocate(global_start(dminfo % nprocs))
+      allocate(global_block_id_arr(partial_global_graph_info % nVerticesTotal))
+      allocate(local_block_id_arr(partial_global_graph_info % nVertices))
+
+      call mpas_timer_start('scotch_total')
+
+      call mpas_log_write('Using LibScotch for graph partitioning')
+
+      ! Count the number of edges (including to ghost cells) in the portion of graph
+      ! owned by the current rank. Each edge is counted twice, once for each vertex,
+      ! with the exception of edges to ghost vertices, which are counted only once.
+      do i=1,partial_global_graph_info % nVertices
+         do j=1,partial_global_graph_info % nAdjacent(i)
+            if (partial_global_graph_info % adjacencyList(j,i) == 0) cycle
+            nLocEdgesGraph = nLocEdgesGraph + 1
+            ! call mpas_log_write('i=$i j=$i adj= $i', intArgs=(/i,j,partial_global_graph_info % adjacencyList(j,i)/) )
+         end do
+      end do
+
+      ! Holds the adjacency array for every local vertex
+      allocate(edgeloctab(nLocEdgesGraph))
+      ! Array of start indices in edgeloctab for each local vertex
+      allocate(vertloctab(partial_global_graph_info % nVertices + 1))
+
+      ! Fill up edgeloctab and vertloctab
+      k = 1
+      do i=1,partial_global_graph_info % nVertices
+         vertloctab(i) = k 
+         do j=1,partial_global_graph_info % nAdjacent(i)
+            if (partial_global_graph_info % adjacencyList(j,i) == 0) cycle
+            edgeloctab(k) = partial_global_graph_info % adjacencyList(j,i)
+            k = k + 1
+         end do
+      end do 
+      vertloctab(partial_global_graph_info % nVertices+1) = nLocEdgesGraph + 1
+
+      ! Duplicate the communicator to be used by Scotch
+      call MPI_Comm_dup(MPI_COMM_WORLD, localcomm, mpi_ierr)
```

Comment:
In general, we can't assume that MPAS owns the entire `MPI_COMM_WORLD` communicator (as might be the case, for example, when MPAS-Atmosphere is operating as a dynamical core in CAM within CESM).
Instead of `MPI_COMM_WORLD`, we can use the communicator found in `domain % dminfo`.

Reply:
Makes sense. Thanks for the catch.